symbolic-base/rust-wrapper/src/prover.rs

```diff
@@ -0,0 +1,888 @@
+
+    v.push(v0);
+
+    for i in 1..n {
+        let prev = v[i - 1];
+        v.push(prev * wInv);
+    }
+    DensePolynomial::from_coefficients_vec(v)
+}
+
+// taken from https://github.com/dusk-network/plonk/blob/master/src/fft/domain.rs
+//
+fn bitreverse(mut n: u32, l: u32) -> u32 {
+    let mut r = 0;
+    for _ in 0..l {
+        r = (r << 1) | (n & 1);
+        n >>= 1;
+    }
+    r
+}
+
+fn serial_fft(a: &mut [ScalarField], omega: ScalarField, log_n: u32) {
+    let n = a.len() as u32;
+    assert_eq!(n, 1 << log_n);
+
+    for k in 0..n {
+        let rk = bitreverse(k, log_n);
+        if k < rk {
+            a.swap(rk as usize, k as usize);
+        }
+    }
+
+    let mut m = 1;
+    for _ in 0..log_n {
+        let w_m = omega.pow(&[(n / (2 * m)) as u64, 0, 0, 0]);
+
+        let mut k = 0;
+        while k < n {
+            let mut w = ScalarField::one();
+            for j in 0..m {
+                let mut t = a[(k + j + m) as usize];
+                t *= &w;
+                let mut tmp = a[(k + j) as usize];
+                tmp -= &t;
+                a[(k + j + m) as usize] = tmp;
+                a[(k + j) as usize] += &t;
+                w.mul_assign(&w_m);
+            }
+
+            k += 2 * m;
+        }
+
+        m *= 2;
+    }
+}
+
+fn polyVecInLagrangeBasis(
+    n: usize,
+    omega: &ScalarField,
+    p: &DensePolynomial<ScalarField>,
+) -> DensePolynomial<ScalarField> {
+    let v = &mut p.coeffs.clone();
+
+    let next_pow2 = v.len().next_power_of_two();
+
+    if next_pow2 > v.len() {
+        v.resize(next_pow2, ScalarField::zero());
+    }
+
+    let norm = ScalarField::from(n as i32).inverse().unwrap();
+
+    let mut norms = Vec::with_capacity(n);
+    for i in 0..n {
+        norms.push(omega.pow([i as u64]) * norm);
+    }
+    v.rotate_right(1);
+
+    let mut result = zip_with(&v, &norms, |a, b| a * b);
+
+    serial_fft(&mut result, *omega, log2(n));
+    result.reverse();
+
+    DensePolynomial::from_coefficients_vec(result)
+}
+
+fn com(gs: &Vec<GAffine>, p: &DensePolynomial<ScalarField>) -> GAffine {
+    let scalars: Vec<_> = p.coeffs.iter().map(|i| i.into_bigint()).collect();
+    let opt_window_size = get_opt_window_size(log2(gs.len()));
+    VariableBaseMSM::multi_scalar_mul_custom(gs, &scalars, opt_window_size, 2048, 256, false).into()
+}
+
+fn compress(pt: &GAffine) -> Vec<u8> {
+    let mut buf = Vec::new();
+    pt.serialize_compressed(&mut buf)
+        .expect("Serialization failed");
+    buf
+}
+
+fn bytes(pt: &ScalarField) -> Vec<u8> {
+    unpack_scalar(*pt)
+}
+
+fn challenge(transcript: &Vec<u8>) -> ScalarField {
+    let digest = blake2b_224(transcript);
+    ScalarField::from_le_bytes_mod_order(&digest)
+}
+
+fn zip_with<A, B, C, F>(a: &[A], b: &[B], mut f: F) -> Vec<C>
+where
+    F: FnMut(&A, &B) -> C,
+{
+    a.iter().zip(b.iter()).map(|(x, y)| f(x, y)).collect()
+}
+
+fn zip_with3<'a, A, B, C, D, F>(v1: &'a [A], v2: &'a [B], v3: &'a [C], f: F) -> Vec<D>
```

Comment:
`&`s have subtyping covariant in lifetimes, so you don't have to constrain arguments to have the same lifetime as far as I can see, so explicit lifetime annotation is not needed